### PR TITLE
#76 - recipe steps

### PR DESCRIPTION
**Pull request by flashback2k14**

### CODE REVIEW

File: server/src/logic/middleware/model-validation.middleware.ts [hunk 1]
```diff
@@ -37,4 +37,39 @@ export class ModelValidationMiddleware {
...
     });
   }
+
+  /**
+   * This validation approache enables also validation of nested objects
+   * @param objToValidate 
+   * @param req 
+   * @param validationGroups 
+   */
+  public validateObject (objToValidate: Validatable, req: Request, validationGroups: ValidationGroup[]): Promise<any> {
```
Comment (flashback2k14) [inline server/src/logic/middleware/model-validation.middleware.ts]: what's the purpose of this function?

File: server/src/logic/validators/recipe-steps-order.validator.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import {
+  registerDecorator, ValidationArguments, ValidationOptions,
+  ValidatorConstraint, ValidatorConstraintInterface
+} from "class-validator";
+import { RecipeStep } from "../../models/database/recipe-step";
+
+@ValidatorConstraint({ name: "RecipeStepsOrder", async: false })
+export class RecipeStepsAreOrderd implements ValidatorConstraintInterface {
+
+  private _sortAsc = (leftNumber: number, rightNumber: number): number => {
```
Comment (flashback2k14) [inline server/src/logic/validators/recipe-steps-order.validator.ts]: this is a unused function.

File: server/src/models/database/recipe-step.ts [hunk 3]
```diff
@@ -1,13 +1,69 @@
...
+    groups: [ValidationGroup.RECIPE]
+  })
+  @MinLength(1, {
+    groups: [ValidationGroup.RECIPE]
+  })
+  public stepDescription: string;
+  public createdDate: number;
+  public lastModifiedDate: number;
+
+  public of (step: RecipeStep): RecipeStep {
```
Comment (flashback2k14) [inline server/src/models/database/recipe-step.ts]: could we call this copy or clone? _of_ confuse me every time i see it. :-)

File: server/src/models/database/validatable.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+export interface Validatable {
```
Comment (flashback2k14) [inline server/src/models/database/validatable.ts]: there is no implementation of this interface. i like interfaces prefixed with an _I_. I don't know why this is bad!?

File: server/src/db/repositories/recipe.repository.ts [hunk 5]
```diff
@@ -11,17 +11,31 @@ export class RecipeRepository {
...
+        return {steps, recipe};
+      });
+  }
+
+  private _addStepsToRecipe = (obj: any) => {
+    obj.recipe.steps = obj.steps;
+    return obj.recipe;
+  }
+
   public getById = (id: number): Promise<Recipe> => {
```
Comment (flashback2k14) [inline server/src/db/repositories/recipe.repository.ts]:
```typescript
private _fetchSteps = (recipe: Recipe) => {
    return this._db.recipeSteps.get(recipe.id)
      .then((steps) => {
        return {steps, recipe};
      })
      .catch(() => {
        return {steps: [], recipe};
      });
  }
```

File: server/src/db/repositories/recipe.repository.ts [hunk 6]
```diff
@@ -11,17 +11,31 @@ export class RecipeRepository {
...
+      .then((obj: any) => this._addStepsToRecipe(obj))
       .catch((error: any) => {
         return ErrorUtils.handleDbNotFound(
           ERRORS.RECIPE_NOT_FOUND, error,
           RecipeRepository.name, "getById(..)", "id", id.toString());
       });
   }
 
-  public getAll = (): Promise<Recipe[]> => {
+  public getAll = (): Promise<Recipe[] | DbError> => {
```
Comment (flashback2k14) [inline server/src/db/repositories/recipe.repository.ts]: this return an empty recipe steps array

File: server/src/db/repositories/recipe.repository.ts (hunk 5 above)
Comment (hetec) [inline server/src/db/repositories/recipe.repository.ts]: Yes that is true! I will fix this

File: server/src/db/repositories/recipe.repository.ts [hunk 7]
```diff
@@ -104,6 +129,7 @@ export class RecipeRepository {
     transformedResult.description = result.hasOwnProperty("description") ? result.description : null;
```
Comment (hetec) [inline server/src/db/repositories/recipe.repository.ts]: OK

File: server/src/db/repositories/recipe-step.repository.ts [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    return this._db.many(query, [recipeId]).then((result) => result.map(this._transformResult));
+  }
+
+  public delete = (recipeId: Number) => {
+    const query: string = `delete from recipe_steps where recipe_steps.recipeId = $1`;
+    return this._db.none(query, [recipeId]);
+  }
+
+  private _transformResult = (result: any): RecipeStep => {
+    const transformedResult: RecipeStep = new RecipeStep();
```
Comment (hetec) [inline server/src/db/repositories/recipe-step.repository.ts]: OK

File: server/src/logic/middleware/model-validation.middleware.ts (hunk 1 above)
Comment (hetec) [inline server/src/logic/middleware/model-validation.middleware.ts]: The original validate method is not able to check nested objects, so the nested recipeSteps were not validated on save.

File: server/src/logic/validators/recipe-steps-order.validator.ts (hunk 2 above)
Comment (hetec) [inline server/src/logic/validators/recipe-steps-order.validator.ts]: Oh yes, maybe I missed to remove it

File: server/src/models/database/recipe-step.ts (hunk 3 above)
Comment (hetec) [inline server/src/models/database/recipe-step.ts]: yes sure if this is more clear for you

File: server/src/models/database/validatable.ts (hunk 4 above)
Comment (hetec) [inline server/src/models/database/validatable.ts]: The recipe and recipeSteps should implement this. I think wanted to still do this.

File: server/src/logic/middleware/model-validation.middleware.ts (hunk 1 above)
Comment (flashback2k14) [inline server/src/logic/middleware/model-validation.middleware.ts]: okay. should we remove the other one?

File: server/src/models/database/validatable.ts (hunk 4 above)
Comment (flashback2k14) [inline server/src/models/database/validatable.ts]: okay

File: server/src/logic/middleware/model-validation.middleware.ts (hunk 1 above)
Comment (hetec) [inline server/src/logic/middleware/model-validation.middleware.ts]: To remove this function we need to adapt the user routes and model to fit the new one.

**Comment by flashback2k14** [comment]
Looks good now. Have you looked into the readme @hetec?

**Pull request merged**